subsys/bluetooth/host/conn.c

```diff
@@ -132,6 +120,54 @@ static struct bt_conn sco_conns[CONFIG_BT_MAX_SCO_CONN];
 #endif /* CONFIG_BT_CLASSIC */
 #endif /* CONFIG_BT_CONN */
 
+#if defined(CONFIG_BT_CONN_TX)
+void frag_destroy(struct net_buf *buf);
+
+/* Storage for fragments (views) into the upper layers' PDUs. */
+/* TODO: remove user-data requirements */
+NET_BUF_POOL_FIXED_DEFINE(fragments, CONFIG_BT_CONN_FRAG_COUNT, 0,
+			  CONFIG_BT_CONN_TX_USER_DATA_SIZE, frag_destroy);
+
+struct frag_md {
+	struct bt_buf_view_meta view_meta;
+};
+struct frag_md frag_md_pool[CONFIG_BT_CONN_FRAG_COUNT];
+
+struct frag_md *get_frag_md(struct net_buf *fragment)
+{
+	return &frag_md_pool[net_buf_id(fragment)];
+}
+
+void bt_tx_irq_raise(void);
+void frag_destroy(struct net_buf *frag)
+{
+	/* allow next view to be allocated (and unlock the parent buf) */
+	bt_buf_destroy_view(frag, &get_frag_md(frag)->view_meta);
+}
+
+static struct net_buf *get_acl_frag(struct net_buf *outside, size_t winsize)
+{
+	struct net_buf *window;
+
+	__ASSERT_NO_MSG(!bt_buf_has_view(outside));
+
+	/* Keeping a ref is the caller's responsibility */
+	window = net_buf_alloc_len(&fragments, 0, K_NO_WAIT);
+	if (!window) {
+		return window;
```

Comment:
I just tested the patch on sample samples/bluetooth/handsfree on my test hw (RT1170 + 2el).

The CPU fault will raise.

I think the possible scenario is:
1. There is more than one buf to be sent, and the length of the first buf is not greater than `conn_mtu(conn)`,
2. In the function `l2cap_br_data_pull`, the first buf will be removed from `br_chan->_pdu_tx_queue`, and the function `raise_data_ready(br_chan)` will be called, because there is more data (secondary buf) to be sent (`&br_chan->_pdu_tx_queue` is not empty).
4. In function `send_buf`, the first buf will allocate a view and put the buf into the H4 send queue.
5. `bt_conn_tx_processor` will call the function `l2cap_br_data_pull` to get the secondary buf, but the first buf has not been completed yet.
6. In function `send_buf`, there is not more view could be allocated for the secondary buf. The buf `frag` is NULL, and it will be sent out via send_acl. And then CPU fault occurs.

Reply:
You're right, I should add an assert that `frag` is not null.
Could you enable `CONFIG_BT_CONN_LOG_LEVEL_DBG=y` so we can verify your hypothesis?

Reply:
Bluetooth initialized
BR/EDR set connectable and discoverable done
[00:00:03.028,000] <dbg> bt_conn: bt_conn_tx_processor: start
[00:00:03.028,000] <dbg> bt_conn: bt_conn_tx_processor: no connection wants to do stuff
[00:00:42.782,000] <dbg> bt_conn: bt_conn_tx_processor: no connection wants to do stuff
--- 6 messages dropped ---
[00:00:42.807,000] <dbg> bt_conn: bt_conn_ref: handle 0 ref 1 -> 2
[00:00:42.807,000] <dbg> bt_conn: bt_conn_set_state: initiating -> connected
--- 2 messages dropped ---
[00:00:42.808,000] <dbg> bt_conn: bt_conn_tx_processor: processing conn 0x200014f0
--- 32 messages dropped ---
[00:00:42.839,000] <dbg> bt_conn: bt_conn_recv: handle 1 len 10 flags 02
[00:00:42.839,000] <dbg> bt_conn: bt_acl_recv: First, len 10 final 6
[00:00:42.839,000] <dbg> bt_conn: bt_acl_recv: Successfully parsed 10 byte L2CAP packet
--- 15 messages dropped ---
[00:00:42.877,000] <dbg> bt_conn: bt_conn_ref: handle 1 ref 2 -> 3
--- 73 messages dropped ---
[00:00:42.886,000] <dbg> bt_conn: tx_complete_work: conn 0x200014f0
--- 20 messages dropped ---
[00:00:42.886,000] <dbg> bt_conn: tx_notify: conn 0x200014f0
--- 6 messages dropped ---
[00:00:42.887,000] <dbg> bt_conn: bt_conn_unref: handle 1 ref 2 -> 1
[00:00:42.887,000] <dbg> bt_conn: tx_complete_work: conn 0x200014f0
[00:00:42.887,000] <dbg> bt_conn: tx_notify: conn 0x200014f0
[00:00:42.887,000] <dbg> bt_conn: tx_notify: tx 0x2000224c cb 0 user_data 0
[00:00:42.887,000] <dbg> bt_conn: tx_free: 0x2000224c
[00:00:42.887,000] <dbg> bt_conn: tx_notify: raise TX IRQ
[00:00:42.887,000] <dbg> bt_conn: bt_conn_tx_processor: start
[00:00:42.887,000] <dbg> bt_conn: bt_conn_tx_processor: no connection wants to do stuff
[00:00:42.890,000] <dbg> bt_conn: bt_conn_ref: handle 1 ref 1 -> 2
[00:00:42.890,000] <dbg> bt_conn: wait_for_tx_work: conn 0x200014f0
[00:00:42.890,000] <dbg> bt_conn: tx_complete_work: conn 0x200014f0
[00:00:42.890,000] <dbg> bt_conn: tx_notify: conn 0x200014f0
[00:00:42.890,000] <dbg> bt_conn: wait_for_tx_work: done
[00:00:42.890,000] <dbg> bt_conn: bt_conn_recv: handle 1 len 20 flags 02
[00:00:42.890,000] <dbg> bt_conn: bt_acl_recv: First, len 20 final 16
[00:00:42.890,000] <dbg> bt_conn: bt_acl_recv: Successfully parsed 20 byte L2CAP packet
[00:00:42.890,000] <dbg> bt_conn: bt_conn_ref: handle 1 ref 2 -> 3
[00:00:42.890,000] <dbg> bt_conn: bt_conn_unref: handle 1 ref 3 -> 2
[00:00:42.890,000] <dbg> bt_conn: bt_conn_unref: handle 1 ref 2 -> 1
[00:00:42.903,000] <dbg> bt_conn: bt_conn_ref: handle 1 ref 1 -> 2
[00:00:42.903,000] <dbg> bt_conn: bt_conn_unref: handle 1 ref 2 -> 1
[00:00:42.903,000] <dbg> bt_conn: bt_conn_ref: handle 1 ref 1 -> 2
[00:00:42.904,000] <dbg> bt_conn: bt_conn_tx_processor: start
[00:00:42.904,000] <dbg> bt_conn: bt_conn_tx_processor: no connection wants to do stuff
[00:00:42.907,000] <dbg> bt_conn: bt_conn_unref: handle 1 ref 2 -> 1
[00:00:42.907,000] <dbg> bt_conn: bt_conn_tx_processor: start
[00:00:42.907,000] <dbg> bt_conn: bt_conn_tx_processor: no connection wants to do stuff
[00:00:43.368,000] <dbg> bt_conn: bt_conn_ref: handle 1 ref 1 -> 2
[00:00:43.368,000] <dbg> bt_conn: bt_conn_tx_processor: start
[00:00:43.368,000] <dbg> bt_conn: bt_conn_tx_processor: no connection wants to do stuff
[00:00:43.371,000] <dbg> bt_conn: bt_conn_unref: handle 1 ref 2 -> 1
[00:00:43.371,000] <dbg> bt_conn: bt_conn_tx_processor: start
[00:00:43.371,000] <dbg> bt_conn: bt_conn_tx_processor: no connection wants to do stuff
[00:00:45.148,000] <dbg> bt_conn: bt_conn_ref: handle 1 ref 1 -> 2
--- 28 messages dropped ---
[00:00:45.199,000] <dbg> bt_conn: bt_conn_tx_processor: start
--- 59 messages dropped ---
[00:00:45.210,000] <dbg> bt_conn: tx_notify: conn 0x200014f0
--- 42 messages dropped ---
[00:00:45.222,000] <dbg> bt_conn: bt_acl_recv: First, len 24 final 20
--- 37 messages dropped ---
[00:00:45.237,000] <dbg> bt_conn: bt_acl_recv: First, len 12 final 8
--- 12 messages dropped ---
[00:00:45.251,000] <dbg> bt_conn: send_buf: conn 0x200014f0 buf 0x20016cac len 16 flags 0x02
--- 49 messages dropped ---
[00:00:45.251,000] <dbg> bt_conn: bt_conn_tx_processor: processing conn 0x200014f0
--- 18 messages dropped ---
[00:00:45.254,000] <dbg> bt_conn: bt_conn_tx_processor: start
--- 41 messages dropped ---
[00:00:45.281,000] <dbg> bt_conn: bt_conn_tx_processor: start
--- 37 messages dropped ---
[00:00:45.295,000] <dbg> bt_conn: tx_complete_work: conn 0x200014f0
--- 3 messages dropped ---
[00:00:45.295,000] <dbg> bt_conn: bt_acl_recv: First, len 24 final 20
[00:00:45.295,000] <dbg> bt_conn: bt_acl_recv: Successfully parsed 24 byte L2CAP packet
--- 38 messages dropped ---
[00:00:45.325,000] <dbg> bt_conn: bt_conn_recv: handle 1 len 24 flags 02
--- 9 messages dropped ---
[00:00:45.326,000] <dbg> bt_conn: bt_conn_tx_processor: processing conn 0x200014f0
--- 30 messages dropped ---
[00:00:45.343,000] <dbg> bt_conn: bt_conn_recv: handle 1 len 12 flags 02
[00:00:45.343,000] <dbg> bt_conn: bt_acl_recv: First, len 12 final 8
--- 8 messages dropped ---
[00:00:45.344,000] <dbg> bt_conn: bt_conn_tx_processor: processing conn 0x200014f0
[00:00:45.344,000] <dbg> bt_conn: bt_conn_tx_processor: pop: cb 0 userdata 0
[00:00:45.344,000] <dbg> bt_conn: bt_conn_tx_processor: TX process: conn 0x200014f0 buf 0x20017458 (last)
--- 48 messages dropped ---
[00:00:45.390,000] <dbg> bt_conn: bt_conn_tx_processor: processing conn 0x200014f0
[00:00:45.390,000] <dbg> bt_conn: bt_conn_tx_processor: no buf returned
[00:00:45.392,000] <dbg> bt_conn: frag_destroy:
[00:00:45.392,000] <dbg> bt_conn: bt_conn_data_ready: DR
[00:00:45.392,000] <dbg> bt_conn: bt_conn_data_ready: raised
[00:00:45.392,000] <dbg> bt_conn: bt_conn_ref: handle 1 ref 2 -> 3
--- 41 messages dropped ---
[00:00:45.441,000] <dbg> bt_conn: bt_conn_tx_processor: pop: cb 0 userdata 0
--- 52 messages dropped ---
[00:00:45.455,000] <dbg> bt_conn: tx_notify: conn 0x200014f0
--- 3 messages dropped ---
[00:00:45.455,000] <dbg> bt_conn: bt_acl_recv: Successfully parsed 31 byte L2CAP packet
--- 39 messages dropped ---
[00:00:45.477,000] <dbg> bt_conn: bt_acl_recv: First, len 12 final 8
[00:00:45.477,000] <dbg> bt_conn: bt_acl_recv: Successfully parsed 12 byte L2CAP packet
[00:00:45.477,000] <dbg> bt_conn: bt_conn_data_ready: DR
--- 7 messages dropped ---
[00:00:45.477,000] <dbg> bt_conn: bt_conn_tx_processor: pop: cb 0 userdata 0
[00:00:45.477,000] <dbg> bt_conn: bt_conn_tx_processor: TX process: conn 0x200014f0 buf 0x20017458 (last)
--- 30 messages dropped ---
[00:00:45.527,000] <dbg> bt_conn: should_stop_tx: 0x200014f0
--- 22 messages dropped ---
[00:00:45.527,000] <dbg> bt_conn: should_stop_tx: No more data for 0x200014f0
[00:00:45.527,000] <dbg> bt_conn: bt_conn_tx_processor: processing conn 0x200014f0
[00:00:45.527,000] <dbg> bt_conn: bt_conn_tx_processor: no buf returned
[00:00:45.529,000] <dbg> bt_conn: frag_destroy:
--- 14 messages dropped ---
[00:00:45.570,000] <dbg> bt_conn: bt_conn_tx_processor: processing conn 0x200014f0
--- 85 messages dropped ---
[00:00:45.580,000] <dbg> bt_conn: bt_acl_recv: First, len 8 final 4
--- 37 messages dropped ---
[00:00:45.593,000] <dbg> bt_conn: bt_acl_recv: First, len 18 final 14
--- 3 messages dropped -

Reply:
> You're right, I should add a check that `frag` is not null. Could you enable `CONFIG_BT_CONN_LOG_LEVEL_DBG=y` so we can verify your hypothesis?

The LOG is attched. But too many messages are dropped, the reference is of little significance.

Reply:
Thanks I just needed the one right before the fault. They do confirm your hypothesis:

>  [00:00:12.168,000] bt_conn: send_buf: conn 0x200017f0 buf 0 len 10921 flags 0x02

Can you try with the latest revision?

Reply:
With latest version, this issue disappeared.

Reply:
perfect, could you please approve the PR?

I'll also try to get a minimal test setup going for Classic as soon as you have added BT support for the NXP devkit in zephyr. That way we can avoid pinging you all the time :)